infra/gcp/ensure-main-project.sh

```diff
@@ -34,249 +34,347 @@ if [ $# != 0 ]; then
-ensure_dns_zone "${PROJECT}" "canary-x-k8s-io" "canary.x-k8s.io"
-ensure_dns_zone "${PROJECT}" "canary-k8s-e2e-com" "canary.k8s-e2e.com"
-
-color 6 "Creating the BigQuery dataset for billing data"
-if ! bq --project_id "${PROJECT}" ls "${BQ_BILLING_DATASET}" >/dev/null 2>&1; then
-    bq --project_id "${PROJECT}" mk "${BQ_BILLING_DATASET}"
-fi
 
-color 6 "Setting BigQuery permissions"
+# Create a GCP service account intended for use by GKE cluster workloads
+# $1: The GCP project hosting the service account (e.g. k8s-infra-foo)
+# $2: The name of the GCP service account (e.g. k8s-infra-doer)
+# $3: The IAM role the service account will have on the project (e.g. roles/foo.bar)
+# $4: The GCP project hosting GKE cluster(s) that will use this service account (e.g. k8s-infra-foo-clusters)
+# $5: The K8s namespace hosting the service account that will use this GCP SA (e.g. my-app-ns)
+#
+# e.g. the above allows pods running as my-app-ns/k8s-infra-doer in clusters in
+#      k8s-infra-foo-clusters to act as k8s-infra-doer@k8s-infra-foo.iam.gserviceaccount.com,
+#      using the permissions granted by roles/foo.bar on k8s-infra-foo
+function ensure_workload_identity_serviceaccount() {
+    if [ $# -lt 5 ] || [ -z "$1" ] || [ -z "$2" ] || [ -z "$3" ] || [ -z "$4" ] || [ -z "$5" ]; then
+        echo "${FUNCNAME[0]}(project, name, role, cluster_project, cluster_namespace) requires 5 arguments" >&2
+        return 1
+    fi
+    local serviceaccount_project="${1}"
+    local serviceaccount_name="${2}"
+    local serviceaccount_email
+    serviceaccount_email="$(svc_acct_email "${serviceaccount_project}" "${serviceaccount_name}")"
+    local serviceaccount_description="${serviceaccount_name}"
+    local serviceaccount_role="${3}"
+    local cluster_project="${4}"
+    local cluster_namespace="${5}"
+
+    color 6 "Ensuring serviceaccount ${serviceaccount_email} exists"
+    ensure_service_account "${serviceaccount_project}" "${serviceaccount_name}" "${serviceaccount_description}"
+
+    color 6 "Empowering ${serviceaccount_email} with role ${serviceaccount_role} in project ${serviceaccount_project}"
+    ensure_project_role_binding \
+        "${serviceaccount_project}" \
+        "serviceAccount:${serviceaccount_email}" \
+        "${serviceaccount_role}"
+
+    local clusters="clusters in project ${cluster_project}"
+    local k8s_sa="KSA ${cluster_namespace}/${serviceaccount_name}"
+    local gcp_sa="GCP SA ${serviceaccount_email}"
+
+    color 6 "Empowering ${clusters} to run workloads as ${k8s_sa} to act as ${gcp_sa}"
+    empower_gke_for_serviceaccount "${cluster_project}" "${cluster_namespace}" "${serviceaccount_email}"
+}
+
+#
+# main functions
+#
 
-# Merge existing permissions with the ones we need to exist.  We merge
-# permissions because:
-#   * The full list is large and has stuff that is inherited listed in it
-#   * All of our other IAM binding logic calls are additive
+function ensure_terraform_state_buckets() {
```

Comment:
Non-blocking comment: IMHO we should empower the owners of each bucket with `roles/viewer`so they can easily run `terraform plan`.

Reply:
Thank you once again for the good catch.  Apparently `roles/viewer` is necessary to get `storage.buckets.list`.

I didn't catch this when testing with my less-privileged account, because it has `audit.viewer` at the org level and thus `roles/viewer` on everything implicitly.

They only other off-the-shelf roles that have this are:
- `roles/storage.admin`
- `roles/viewer`
- `roles/editor`
- `roles/owner`

I'm tempted to suggest `storage.admin`, let me look real quick

Reply:
So it looks like we can't set any of the primitive roles just on the bucket
```
$ gsutil iam ch "user:spiffxp@google.com:roles/viewer" gs://k8s-infra-clusters-terraform
BadRequestException: 400 Role (roles/viewer) does not exist in the resource's hierarchy.
```

Which leaves use with `roles/viewer` on the project, or `roles/storage.admin` on the bucket as the quick fix here.  I'll add to the project for now

Reply:
Opened https://github.com/kubernetes/k8s.io/pull/2005 to address